### PR TITLE
Cart Tax Applicator

### PR DESCRIPTION
**Pull request by dallendalton**
Continuing the work of refactor tax is calculated and applied to a cart, this PR includes the functionality for apply tax to the cart.

### CODE REVIEW

File: includes/TaxCalculation/class-cart-tax-applicator.php [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+
+	/**
+	 * Builds a WooCommerce structured tax rate.
+	 *
+	 * @param float  $rate Tax rate.
+	 * @param string $shipping_taxable yes if shipping is taxable, no if not.
+	 *
+	 * @return array[]
+	 */
+	private function build_woocoommerce_rate( $rate, $shipping_taxable = 'no' ): array {
```
Comment (poops) [inline includes/TaxCalculation/class-cart-tax-applicator.php:134]: Typo in "woocoommerce"

File: includes/TaxCalculation/class-cart-tax-applicator.php [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+	 * @param array  $item item in cart.
+	 * @param array  $wc_rate WooCommerce structure rate.
+	 *
+	 * @return array
+	 */
+	private function apply_line_subtotal_tax( $item_key, $item, $wc_rate ): array {
+		$item_subtotal  = wc_add_number_precision( $item['data']->get_price() * $item['quantity'], false );
+		$subtotal_taxes = WC_Tax::calc_tax( $item_subtotal, $wc_rate );
+		$subtotal_tax   = array_sum( array_map( array( $this, 'round_line_tax' ), $subtotal_taxes ) );
+		$this->cart->cart_contents[ $item_key ]['line_tax_data']     = array( 'subtotal' => wc_remove_number_precision_deep( $subtotal_taxes ) );
```
Comment (poops) [inline includes/TaxCalculation/class-cart-tax-applicator.php:158]: Do you think there's any benefit in having the `line_tax_data` array instantiated if needed in this method & `apply_line_total_tax`, so the order of the function calls don't matter?

**Review by poops — APPROVED** [review]
Minor comments, but LGTM

File: includes/TaxCalculation/class-cart-tax-applicator.php (hunk 2 above)
Comment (sethobey) [reply includes/TaxCalculation/class-cart-tax-applicator.php:158]: I see how it is procedurally safe as-is, but I think @poops has a point.

If it was something like:

`$this->cart->cart_contents[ $item_key ]['line_tax_data']['subtotal'] = wc_remove_number_precision_deep( $subtotal_taxes );
 `

as you do on line 175 below, then if these methods were called in a different order, it wouldn't be destructive to the 'total' value set in `apply_line_total_tax`

**Review by sethobey — APPROVED** [review]

File: includes/TaxCalculation/class-cart-tax-applicator.php (hunk 1 above)
Comment (dallendalton) [reply includes/TaxCalculation/class-cart-tax-applicator.php]: Good catch, fixed!

File: includes/TaxCalculation/class-cart-tax-applicator.php (hunk 2 above)
Comment (dallendalton) [reply includes/TaxCalculation/class-cart-tax-applicator.php]: I agree this is a better way and have fixed it. Thanks!

**Pull request merged**